pubpub/pubpub
# fix: Upward permission checking

`scopeGet` handles trickle down permissions. So, if a user is a Community member,

```
Community [isMember]
	Collection []
		Pub []
```

their community-scope permission will be applied to all children elements.

If however, a user is a Pub member,

```
Community []
	Collection []
		Pub [isMember]
```

when viewing the Community dashboard, Collection dashboard, a Page, or a User Profile (the last two defaulting to Community-scope), Pub-level permissions are not calculated and fed upward. The user will not see the Pub in any of these views even though they have permissions to the pub.

`scopeGet` was built around 'can you see this thing' - not 'can you see this thing or any of its children'. This becomes an issue for 'overview' views - any view that's showing an overview of content items, rather than focusing on the scope-level target. So, this isn't an issue (that I can see) for API routes or Pub views, as those are always focused on the specific `activeTarget` element.

Content (Collections or Pubs) is stripped out using sanitize functions (pubSanitize.js and collectionSanitize.js). My thinking is that we have to add a check in each sanitize function that says 'if !canView from activeScope permissions, check for individual element permission'.

Querying scopeGet for each pub and then filtering based on that led to 16s sanitize times locally (2ms originally), ~2s when on a production server. So - that was a dead-end performance wise.

Adding Members to pubs and filtering on that led to no discernible change in query time. (< 1ms)
Adding Members to collections and filtering on that led to no discernible change in query time. (< 1ms)

When we fully implement PublicPermissions at Collection and Community level, the same sort of upward-permission-checks will be necessary.

_Test Plan:_
- Use the Review App deployment below and log in with a Community Admin account (user1) in one window and a non-admin account (user2) in an incognito window
- Assign user2 as a Pub-level Member and verify that they can view the Pub from the Community Dashboard Overview pane even though they only have pub-level permissions
- Verify the same for viewing a Pub on a Page, on the User's profile, and at the Collection Dashboard Overview pane 
- Verify that the above is not true on the same routes currently live on duqduq.org
